Documentation of TLSSocket behavior on AUTH_FAILURE

michalpasztamobica opened the pull request:
### Description

It is not possible to store TLS certificate after opening a TLSSocket, due to mbedTLS limitation. We update the documentation to warn users and explain the reason.
(Internal JIRA with the full discussion is ONME-4121)

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [x] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@SeppoTakalo 
@VeijoPesonen

michalpasztamobica committed:
Documentation of TLSSocket behavior on AUTH_FAILURE

kjbracey commented:
Surely this is nothing to do with "AUTH_FAILURE"? It's a general restriction that the socket can only be used for one connection (attempt).

That limitation is not that unusual for connection-type sockets in general, but may be worth noting here.

(POSIX says: "If connect() fails, the state of the socket is unspecified. Conforming applications should close the file descriptor and create a new socket before attempting to reconnect.")

0xc0170 commented:
Note: Travis failure is fixed on master (https://github.com/ARMmbed/mbed-os/pull/9391) , if you can rebase to resolve the error

michalpasztamobica commented:
@0xc0170 , thanks, I rebased.
@kjbracey-arm , thank you for your remark. I modified the comment. I guess if we warn users that after close() they need to create the socket from scratch, we can also warn them that they should do so in case connect() fails. Is the current comment content acceptable?

SeppoTakalo commented on features/netsocket/TLSSocket.h:
```diff
@@ -75,6 +75,9 @@ class TLSSocket : public TLSSocketWrapper {
      *  a domain name or an IP address and a port.
      *
+     *  @note: In case connect() returns an error, the state of the socket is
+     *  unspecified. A new socket should be created before reconnecting.
```
`unspecified` is not quite accurate.

See the documentation for `mbedtls_ssl_handshake()`
https://tls.mbed.org/api/ssl_8h.html#a4a37e497cd08c896870a42b1b618186e

We should reflect that in our documentation. So if this function returns failures other than what we specify, user should stop using the TLS context and free it. We don't support the `mbedtls_ssl_session_reset()` but maybe we should.

SeppoTakalo replied:
It might be hard to reflect, but basically we could try the `connect()` as long as it is the TCP session that fails the connection.

Basically, if SSL handshake fails, we return `NSAPI_ERROR_AUTH_FAILURE` and after that, the socket cannot be used anymore.

michalpasztamobica replied:
> Basically, if SSL handshake fails, we return NSAPI_ERROR_AUTH_FAILURE

I noticed we sometimes return PARAMETER_ERROR instead, but I saw no reason to do that and I modified the return value. Please take a look and if you agree with that change, I will amend the test's expectations (the tests documentation just says "must return an error", without specifying the exact error code, so that's fine).
If we always return AUTH_ERROR on ssl-related errors we can safaely tell the users that if they see this error, then they must get a new socket before retrying (which means they don't have to do that if they just want to retry the TCP connection).
Is that ok?

SeppoTakalo commented:
Do we actually need to reflect this `connect()` thing in the main Socket.h?

We seem to allow `TCPSocket::connect()` to be retried, but that is not necessary always the case. Therefore I propose that we amend the `Socket::connect()` documentation as well.

SeppoTakalo commented on features/netsocket/TLSSocket.h:
```diff
@@ -75,6 +75,9 @@ class TLSSocket : public TLSSocketWrapper {
      *  a domain name or an IP address and a port.
      *
+     *  @note: In case connect() returns an NSAPI_ERROR_AUTH_FAILURE,
+     *  a new socket should be created before reconnecting.
```
Should this sentence reflect more about this socket, instead of how to proceed?

For example: "In case connect() returns error, you must stop using the socket and free it either by calling its close() function or destroying it"

michalpasztamobica replied:
Shouldn't we stick to  NSAPI_ERROR_AUTH_FAILURE? Other errors do not really render the socket useless. It is just a recommendation in Socket.h to create a new one... 
Otherwise, I agree. Is it okay to be a little more formal, like so: "In case connect() returns NSAPI_ERROR_AUTH_FAILURE, the socket must be freed either by calling close() or destroying it"?

SeppoTakalo replied:
Yes. That makes sense.

0xc0170 commented:
@michalpasztamobica All reviews addressed (we can see a new commit here)

michalpasztamobica commented:
I think they are addressed, although I would prefer to see an explicit OK from @SeppoTakalo  regarding the error code returns ;-)
I checked - no test code modification are needed because we don't check for a particular error code (as the documentation suggests).

michalpasztamobica commented:
@0xc0170 , Travis passed and I think that the builds which were failing so far should now pass (as they don't really compile anything any more). Could you trigger the CI, please?

SeppoTakalo reviewed (approved):
Looks good

0xc0170 commented:
Ready for CI, please rebase first to resolve conflicts (documentation update was merged recently)

michalpasztamobica commented:
Rebased and conflicts resolved.

0xc0170 commented:
CI started

michalpasztamobica commented:
Sorry, I checked Icetea tests and greentea tests, but forgot about the unit tests - latest push fixes the expected error return values.

0xc0170 commented:
CI restarted

michalpasztamobica commented:
Now the CI pointed out that I should also modify the DTLSSocketWrapper tests. I hope nothing more will come up...

0xc0170 commented:
CI restarted

michalpasztamobica commented:
I doubt my changes are causing the dynamic-memory-usage failure as I can see the same is happening in @VeijoPesonen 's [Pull Request](https://github.com/ARMmbed/mbed-os/pull/9421)...

0xc0170 commented:
Yes, test team is investigating.

michalpasztamobica commented:
I can see that 4 out of 5 failing greentea tests are not even related to ipcore.
The error seems to be that the resource NUCLEO_F429ZI wasn't allocated correctly:
```
[1548175529.18][GLRM]079602210542680C3A06FEA7 - Received response
[1548175529.18][GLRM]079602210542680C3A06FEA7 - 409 Conflict

[1548175532.48][HTST][WRN] stopped to consume events due to __notify_sync_failed event
```
@0xc0170 , is this something I should worry about?

0xc0170 commented:
CI restarted

adbridge commented:
I can't seem to patch this with either git am or git cherry-pick so I am bumping it to 5.12

SeppoTakalo commented:
5.12 is fine
